module/zfs/dsl_dataset.c

```diff
@@ -727,26 +802,41 @@ dsl_dataset_disown(dsl_dataset_t *ds, void *tag)
 	ASSERT(ds->ds_dbuf != NULL);
 
 	mutex_enter(&ds->ds_lock);
+	if (ds->ds_dir && ds->ds_dir->dd_crypto_obj) {
+		(void) spa_keystore_remove_mapping(ds->ds_dir->dd_pool->dp_spa,
+		    ds, ds);
+	}
 	ds->ds_owner = NULL;
 	mutex_exit(&ds->ds_lock);
 	dsl_dataset_long_rele(ds, tag);
 	dsl_dataset_rele(ds, tag);
 }
 
-boolean_t
-dsl_dataset_tryown(dsl_dataset_t *ds, void *tag)
+int
+dsl_dataset_tryown(dsl_dataset_t *ds, void *tag, boolean_t key_required)
 {
-	boolean_t gotit = FALSE;
+	int ret = 0;
+	spa_t *spa = ds->ds_dir->dd_pool->dp_spa;
+	uint64_t dckobj = ds->ds_dir->dd_crypto_obj;
 
 	ASSERT(dsl_pool_config_held(ds->ds_dir->dd_pool));
 	mutex_enter(&ds->ds_lock);
 	if (ds->ds_owner == NULL && !DS_IS_INCONSISTENT(ds)) {
+		if (dckobj != 0 && key_required) {
+			ret = spa_keystore_create_mapping(spa, ds, ds);
+			if (ret) {
```

Comment:
Do you want me to make this change everywhere? I have always used `if (ret) {` and so it occurs in a lot of places throughout this patch.